Enable matrices in HLSL

pow2clk opened the pull request:
HLSL needs matrix support. This allows matrices when the target language is HLSL and defines a matrix template alias that allows the short forms of matrix types to be defined in typedefs in a default header. Makes some changes to how matrices are printed in HLSL policy. This required a tweak to how existing matrices are printed in diagnostics that is more consistent with other types.

These matrix types will function exactly as the clang matrix extension dictates. Alterations to that behavior both specific to HLSL and also potentially expanding on the matrix extension will follow.

fixes #109839

Commit:
Enable matrices in HLSL

HLSL needs matrix support. This allows matrices when the target
language is HLSL and defines a matrix template alias that allows
the short forms of matrix types to be defined in typedefs in a
default header. Makes some changes to how matrices are printed
in HLSL policy. This required a tweak to how existing matrices are
printed in diagnostics that is more consistent with other types.

These matrix types will function exactly as the clang matrix extension
dictates. Alterations to that behavior both specific to HLSL and
also potentially expanding on the matrix extension will follow.

fixes #109839

Commit:
clang-format

llvm-beanz commented on clang/lib/AST/TypePrinter.cpp:
```diff
@@ -852,34 +852,50 @@ void TypePrinter::printExtVectorAfter(const ExtVectorType *T, raw_ostream &OS) {
   printBefore(T->getElementType(), OS);
-  OS << " __attribute__((matrix_type(";
-  OS << T->getNumRows() << ", " << T->getNumColumns();
-  OS << ")))";
```
This change seems wrong to me. IIUC this changes something like `float  __attribute__((matrix_type(4, 4)))*` to `float * __attribute__((matrix_type(4, 4)))`, which would mean the element type is `float*` rather than `float`.

pow2clk replied:
See my response to Florian [here](https://github.com/llvm/llvm-project/pull/111415/files#r1796090424)

fhahn replied:
may be simpler to read to duplicate some code but have the HSL and C++ matrixes types printed completely separately?

farzonl commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -1381,7 +1381,7 @@ void ASTContext::InitBuiltinTypes(const TargetInfo &Target,
     InitBuiltinType(ArraySectionTy, BuiltinType::ArraySection);
   }
-  if (LangOpts.MatrixTypes)
+  if (LangOpts.MatrixTypes || LangOpts.HLSL)
```
If this is the only line this will be needed then feel free to ignore this comment. 

My question is  could there be places where `LangOpts.MatrixTypes` could diverge from HLSL? 

If so instead is it possible for `LangOpts.HLSL` to turn on `LangOpts.MatrixTypes`?

llvm-beanz replied:
We can add `MatrixTypes` to the definition of HLSL in LangStandards.def, which is probably a safer approach to this.

pow2clk replied:
I'll look into that. I didn't want to enable matrices by default because I didn't really want to allow the matrix attribute syntax in HLSL.

pow2clk replied:
I've done this two ways now:
https://github.com/llvm/llvm-project/pull/111415/commits/ae1a58be4272bd652fb9891e7670f05574b9eb7e just enables matrices when HLSL is enabled
https://github.com/llvm/llvm-project/pull/111415/commits/eeb31651eeff04a612829c7858955772ba2b2b6e attempts to use LangStandards and make it dependent on HLSL versions

I confess I don't know why the second is preferable to the first. If I've misunderstood @llvm-beanz's suggestion, please let me know.

I'm actually unclear on how any of this prevents the problem @farzonl brought up. If what MatrixTypes represents changes so that it no longer matches what HLSL wants, then those changes will seep in without our notice. Given that the design discussion we had about this rejected the notion that we'd forge our own matrix type that inherited from the clang type, I don't see any way around that problem unless we pay attention to its evolution and weigh in when needed. Not that I'm saying that's a bad idea regardless.

All that the above changes alter is that now matrices can be declared using the clang extension which will sidestep at least the size restrictions we have not yet applied, but intend to.

I'm unconvinced this is an improvement, but it was easy enough to do that I thought I'd give us more concrete options to discuss.

pow2clk commented on clang/test/Sema/matrix-type-operators.c:
```diff
@@ -17,12 +17,12 @@ void add(sx10x10_t a, sx5x10_t b, sx10x5_t c) {
   a = b + &c;
-  // expected-error@-1 {{invalid operands to binary expression ('sx5x10_t' (aka 'float __attribute__((matrix_type(5, 10)))') and 'sx10x5_t *' (aka 'float  __attribute__((matrix_type(10, 5)))*'))}}
-  // expected-error@-2 {{casting 'sx10x5_t *' (aka 'float  __attribute__((matrix_type(10, 5)))*') to incompatible type 'float'}}
+  // expected-error@-1 {{invalid operands to binary expression ('sx5x10_t' (aka 'float __attribute__((matrix_type(5, 10)))') and 'sx10x5_t *' (aka 'float * __attribute__((matrix_type(10, 5)))'))}}
```
hi @fhahn!

Thanks for taking a look at this!

It was intentional, but this is the consequence of the change I commented on [here](https://github.com/llvm/llvm-project/pull/111415/files#r1790670980) hoping you'd weigh in on. I made the change this way in TypePrinter because it was more convenient to insert the HLSL printing style that way, but I can maintain the previous behavior. The reason I thought that perhaps the alternate position of `*` might be correct or preferable is because it is the way that vectors are printed with an admittedly contrived and double erroneous [example](https://godbolt.org/z/Wjzzx7sov).

At any rate, we can defer the discussion of whether the vector printing should conform with the matrix printing or vice versa for later. For now, I'll implement the printing so that it maintains previous behavior.

pow2clk replied:
Another oddity in how pointers to matrices were printed before was a double space between the type and the attribute. It's minor, but it suggested something was missing there.

llvm-beanz replied:
The new printing is illegal. You've changed it from a matrix of `float` to a matrix of `float*`. The type printer shouldn't be printing types in a way that changes their meaning.

pow2clk replied:
Right, then perhaps we should look into how vectors are printed.

pow2clk replied:
I created an issue for this: https://github.com/llvm/llvm-project/issues/114868 It's not really related to HLSL as we don't have pointers or references (yet). I have a tentative fix, but I'd like to see if anyone has thoughts on the further impact it will have on existing type printing.

Commit:
Revert change to how existing matrix types are printed

There is an inconsistency in how matrix and vector pointers and references are printed at present.

Also back out a change that's probably better independent than as a rider

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_matrix_pr

Commit:
Enable matrix types in select HLSL standards

Instead of adding an HLSL component to the checks for matrix types,
this just enables them whenever an HLSL version that has that explicitly set is chosen

Commit:
Enable matrices when HLSL is enabled

Commit:
Purge short and long from tests

Future uncertain. Better to test with native HLSL types anyway

Commit:
Revert "Enable matrix types in select HLSL standards"

This reverts commit eeb31651eeff04a612829c7858955772ba2b2b6e.

Using LangOpts->MatrixTypes for both the default value and the keypath meant that when the lang opts were regenerated, it found them the same and determined that replicating the enable-matrix flag wasn't necessary. Using the HLSL keypath differentiates them so that it will know that they need to be regenerated.

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_matrix_pr

Commit:
correct sizes of one test to conform with HLSL limits

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_matrix_pr

llvm-beanz commented on clang/test/CodeGenCXX/matrix-type-operators.cpp:
```diff
@@ -1,7 +1,5 @@
 // RUN: %clang_cc1 -O0 -fenable-matrix -triple x86_64-apple-darwin %s -emit-llvm -disable-llvm-passes -o - -std=c++11 | FileCheck --check-prefixes=CHECK,NOOPT %s
 // RUN: %clang_cc1 -O1 -fenable-matrix -triple x86_64-apple-darwin %s -emit-llvm -disable-llvm-passes -o - -std=c++11 | FileCheck --check-prefixes=CHECK,OPT %s
-typedef double dx5x5_t __attribute__((matrix_type(5, 5)));
-using fx2x3_t = float __attribute__((matrix_type(2, 3)));
```
Are the changes to this file required?

pow2clk replied:
Not really, but I noted that they weren't being used. I think the other changes to this file are useful as well, but I'm happy to remove them to expedite things.

llvm-beanz commented on clang/lib/AST/TypePrinter.cpp:
```diff
@@ -852,34 +852,51 @@ void TypePrinter::printExtVectorAfter(const ExtVectorType *T, raw_ostream &OS) {
+  if (Policy.UseHLSLTypes)
+    OS << ", ";
+  else
+    OS << " __attribute__((matrix_type(";
```
You still have this in the "after" function. That will print this after pointer annotations which changes its meaning incorrectly.

I'm guessing there isn't an existing test case for this, otherwise this change would have broken something.

cc: @fhahn

pow2clk replied:
Yeah. I guess I missed this one. AFAICT, DependentSizedMatrices are never printed.

Commit:
revert cpp test changes and revise matrix printing

To make this perfectly limited to HLSL, revert matrix cpp test

Revise matrix printing so that should they have pointers or references in HLSL, they will be printed properly as well

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_matrix_pr

Commit:
make a few changes to matrix tests needed for DXIL compat

Now that DXIL uses the itanium C++ ABI, it can use some of the tests that were SPIRV only before

Commit:
Merge remote-tracking branch 'refs/remotes/origin/main' into hlsl_matrix_pr

Commit:
update matrix tests to new splat disasm
